lucene/sandbox/src/test/org/apache/lucene/search/TestLargeNumHitsTopDocsCollector.java

```diff
@@ -0,0 +1,141 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.lucene.search;
+
+import java.io.IOException;
+
+import org.apache.lucene.document.Document;
+import org.apache.lucene.document.Field;
+import org.apache.lucene.index.IndexReader;
+import org.apache.lucene.index.RandomIndexWriter;
+import org.apache.lucene.index.Term;
+import org.apache.lucene.store.Directory;
+import org.apache.lucene.util.LuceneTestCase;
+
+public class TestLargeNumHitsTopDocsCollector extends LuceneTestCase {
+  private Directory dir;
+  private IndexReader reader;
+  private final Query testQuery = new BooleanQuery.Builder()
+      .add(new TermQuery(new Term("field", "5")), BooleanClause.Occur.SHOULD)
+      .add(new MatchAllDocsQuery(), BooleanClause.Occur.SHOULD)
+      .build();
+
+  @Override
+  public void setUp() throws Exception {
+    super.setUp();
+    dir = newDirectory();
+    RandomIndexWriter writer = new RandomIndexWriter(random(), dir);
+    for (int i = 0; i < 200_000; i++) {
+      if (random().nextBoolean()) {
+        Document doc = new Document();
+        doc.add(newStringField("field", "5", Field.Store.NO));
+        writer.addDocument(doc);
+      }
+      else {
+        writer.addDocument(new Document());
+      }
+    }
+    reader = writer.getReader();
+    writer.close();
+  }
+
+  @Override
+  public void tearDown() throws Exception {
+    reader.close();
+    dir.close();
+    dir = null;
+    super.tearDown();
+  }
+  public void testLargeNumAndSparseHits() throws Exception {
+    runNumHits(100_000);
+  }
+
+  public void testSingleNumHit() throws Exception {
+    runNumHits(1);
+  }
+
+  public void testLowNumberOfHits() throws Exception {
+    runNumHits(25);
+  }
+
+  public void testIllegalArguments() throws IOException {
+    Query q = new MatchAllDocsQuery();
+    IndexSearcher searcher = newSearcher(reader);
+    LargeNumHitsTopDocsCollector largeCollector = new LargeNumHitsTopDocsCollector(15);
+    TopScoreDocCollector regularCollector = TopScoreDocCollector.create(15, null, Integer.MAX_VALUE);
+
+    searcher.search(q, largeCollector);
+    searcher.search(q, regularCollector);
+
+    assertEquals(largeCollector.totalHits, regularCollector.totalHits);
+
+    IllegalArgumentException expected  = expectThrows(IllegalArgumentException.class, () -> {
+      largeCollector.topDocs(350_000);
+    });
+
+    assertTrue(expected.getMessage().contains("Incorrect number of hits requested"));
+  }
+
+  public void testNoPQBuild() throws IOException {
+    Query q = new MatchAllDocsQuery();
+    IndexSearcher searcher = newSearcher(reader);
+    LargeNumHitsTopDocsCollector largeCollector = new LargeNumHitsTopDocsCollector(250_000);
+    TopScoreDocCollector regularCollector = TopScoreDocCollector.create(250_000, null, Integer.MAX_VALUE);
+
+    searcher.search(q, largeCollector);
+    searcher.search(q, regularCollector);
```

Comment:
can you use testQuery instead to make sure that the list is sorted before creating top docs?